src/atomic.cr

```diff
@@ -31,90 +35,97 @@ struct Atomic(T)
   # atomic.get                   # => 3
   # ```
   def compare_and_set(cmp : T, new : T) : {T, Bool}
-    # Check if it's a nilable reference type
-    {% if T.union? && T.union_types.all? { |t| t == Nil || t < Reference } %}
-      # If so, use addresses because LLVM < 3.9 doesn't support cmpxchg with pointers
-      address, success = Ops.cmpxchg(pointerof(@value).as(LibC::SizeT*), LibC::SizeT.new(cmp.as(T).object_id), LibC::SizeT.new(new.as(T).object_id), :sequentially_consistent, :sequentially_consistent)
-      {address == 0 ? nil : Pointer(T).new(address).as(T), success}
-      # Check if it's a reference type
-    {% elsif T < Reference %}
-      # Use addresses again (but this can't return nil)
-      address, success = Ops.cmpxchg(pointerof(@value).as(LibC::SizeT*), LibC::SizeT.new(cmp.as(T).object_id), LibC::SizeT.new(new.as(T).object_id), :sequentially_consistent, :sequentially_consistent)
-      {Pointer(T).new(address).as(T), success}
-    {% else %}
-      # Otherwise, this is an integer type
-      Ops.cmpxchg(pointerof(@value), cmp, new, :sequentially_consistent, :sequentially_consistent)
-    {% end %}
+    Ops.cmpxchg(pointerof(@value), cmp.as(T), new.as(T), :sequentially_consistent, :sequentially_consistent)
   end
 
   # Performs `atomic_value &+= value`. Returns the old value.
   #
+  # `T` cannot contain any reference types.
```

Comment:
I'd suggest adding a `NOTE` admonition.

ditto elsewhere

Reply:
Can you elaborate why? This seems like a very basic part of the description, rather than a callout.

Reply:
It's a hard requirement, thus IMO deserves visibility.

Reply:
But that's not exactly what `NOTE` admonitions are for.

Reply:
It's hard to use admonitions or agree on what they should mean. I *think* NOTE: should be the same as a footnote, but it's hard to stick to this definition because the reality is that the current look increases visibility rather than decrease it.

Reply:
Oh, I didn't know they're defined (not in the docs they aren't at least) ;)